Move funclets invocation to code managers

janvorli opened the pull request:
This change moves exception handling funclets invocation to the `ICodeManager` derived classes so that it can be done differently for JITted/crossgened code and interpreted code.

janvorli committed:
Move funclets invocation to code managers

This change moves exception handling funclets invocation to the
ICodeManager so that it can be done differently for JITted/crossgened
code and interpreted code.

janvorli commented:
cc: @filipnavara

filipnavara commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3435,20 +3406,10 @@ extern "C" void * QCALLTYPE CallCatchFunclet(QCall::ObjectHandleOnStack exceptio
+        dwResumePC = exInfo->m_frameIter.m_crawl.GetCodeManager()->CallFunclet(throwable, pHandlerIP, pvRegDisplay, exInfo, false /* isFilterFunclet */);
 
+#ifdef USE_FUNCLET_CALL_HELPER
         FixContext(pvRegDisplay->pCurrentContext);
```
It would make sense to move the `FixContext` logic inside `EECodeManager::CallFunclet`. Afterall I don't expect it to be relevant for interpreter (ignoring the fact that we don't support interpreter on x86) and it only calls back into the code manager anyway.

janvorli replied:
I've actually had it there, but moved it back. It is used for catch funclet only, so I would have to add another flag to the signature to distinguish catch. The context fixing is only relevant to the resuming after catch and that happens here, not in the code manager. I would actually prefer replacing the fixcontext here by its body and ifdef it on x86, since the function does nothing for all other architectures.

janvorli commented:
There is some Linux problem, I am investigating.

janvorli committed:
Fix m_csfEHClause - it needs to be SP of the direct funclet caller

janvorli committed:
Fix build for platforms with USE_FUNCLET_CALL_HELPER

I've missed the fact that the macro is defined in the
exceptionhandling.cpp only, so the move to eetwain.cpp disabled this
feature.

janvorli committed:
Fix WASM build

jkotas reviewed (changes requested):
Unix test failures look real

janvorli commented:
> Unix test failures look real

@jkotas Yes, I've been looking into that whole day and finally found the culprit. The C++ compiler optimized the call to the actual funclet from the CallFunclet as a tail call, so the m_csfEHClause wasn't the caller SP. The StackFrameIterator depends on that. I have fixed that by disabling optimizations for the CallFunclet on Windows / disabling tail calling on Unix. I am running the tests now and will update the PR soon.

janvorli committed:
Fix random crashes due to tail call optimization

jkotas commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2124,7 +2128,163 @@ void EECodeManager::LeaveCatch(GCInfoToken gcInfoToken,
+#else
+    HandlerFn* pfnHandler = (HandlerFn*)pHandler;
+
+    pExInfo->m_csfEHClause = CallerStackFrame((UINT_PTR)GetCurrentSP());
```
Can this line be in the `#else` of `#ifdef USE_FUNCLET_CALL_HELPER`?

jkotas replied:
(It would be nice to update x64 to use the asm helper too at some point, so we should keep all workarounds related to that under the !asm helper ifdef.

Switching x64 to asm helper should improve code size of the EH codegen and remove unnecessary differences between architectures.

jkotas commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2124,7 +2128,163 @@ void EECodeManager::LeaveCatch(GCInfoToken gcInfoToken,
+// NOTE: This function must be prevented from calling the actual funclet via a tail call to ensure
+// that the m_csfEHClause is really set to what is a SP of the caller frame of the funclet. The
+// StackFrameIterator relies on this.
+#ifdef _MSC_VER
```
Can this whole workaround be under `#ifndef USE_FUNCLET_CALL_HELPER`?

janvorli committed:
Fix build with GCC and reflect feedback

am11 commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2220,11 +2220,15 @@ static UINT_PTR GetEstablisherFrame(REGDISPLAY* pvRegDisplay, ExInfo* exInfo)
-__attribute__((disable_tail_calls))
+#elif defined(__clang__)
+[[clang::disable_tail_calls]]
+#elif defined(__GNUC__)
```
nit: just `#else` might be better for default (O0) case.

janvorli committed:
Address feedback

jkotas commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2124,7 +2128,169 @@ void EECodeManager::LeaveCatch(GCInfoToken gcInfoToken,
+// * Catch funclet: address to resume at after the catch returns
+// * Finally funclet: unused
+// * Filter funclet: result of the filter funclet (EXCEPTION_CONTINUE_SEARCH (0) or EXCEPTION_EXECUTE_HANDLER (1))
+//
+// NOTE: This function must be prevented from calling the actual funclet via a tail call to ensure
+// that the m_csfEHClause is really set to what is a SP of the caller frame of the funclet. The
+// StackFrameIterator relies on this.
+#ifndef USE_FUNCLET_CALL_HELPER
```
```suggestion
#ifndef USE_FUNCLET_CALL_HELPER
// NOTE: This function must be prevented from calling the actual funclet via a tail call to ensure
// that the m_csfEHClause is really set to what is a SP of the caller frame of the funclet. The
// StackFrameIterator relies on this.
```

jkotas committed:
Update src/coreclr/vm/eetwain.cpp

jkotas commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2124,7 +2128,167 @@ void EECodeManager::LeaveCatch(GCInfoToken gcInfoToken,
+    _ASSERTE(!"CallFunclet for WASM not implemented yet");
+#else
+    HandlerFn* pfnHandler = (HandlerFn*)pHandler;
+
+    pExInfo->m_csfEHClause = CallerStackFrame((UINT_PTR)GetCurrentSP());
```
```suggestion
```

jkotas commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2124,7 +2128,167 @@ void EECodeManager::LeaveCatch(GCInfoToken gcInfoToken,
+                                 pFuncletCallerSP);
+    }
+#else
+    UINT_PTR establisherFrame = GetEstablisherFrame(pRD, pExInfo);
```
```suggestion
    pExInfo->m_csfEHClause = CallerStackFrame((UINT_PTR)GetCurrentSP());
    
    UINT_PTR establisherFrame = GetEstablisherFrame(pRD, pExInfo);
```

jkotas replied:
Addressing my remaining suggestion

janvorli replied:
I am not sure what this suggestion is - the diff shows it in a very confusing way.

jkotas replied:
The suggestion is to move `pExInfo->m_csfEHClause = CallerStackFrame((UINT_PTR)GetCurrentSP());` under `!USE_FUNCLET_CALL_HELPER`

More context: https://github.com/dotnet/runtime/pull/114180#discussion_r2029400483

janvorli replied:
No, this needs to be done in both cases

janvorli replied:
Hmm, actually, it doesn't. My mistake.

janvorli committed:
Another feedback

jkotas committed:
Merge branch 'main' into move-funclet-calls-to-code-manager
